### PR TITLE
V0.2 backend dynamic node support

### PR DESCRIPTION
**Pull request by chaosddp**
# Description

Main changes:
1. use an enum like class to specified attribute,
2. support more attribute data type for both 2 backends.
3. use a simple string template to generate dup codes before compiling.
4. a backend write with c++ to support dynamic attributes
5. add 2 attribute parameters,
    const: same as normal, but not copied into snapshot list, can be queried even without snapshot taking (must specified a tick for padding)
    list: dynamic length attribute, support remove, append, clear, resize and slice accessing.

    An attribute cannot both be const and list, it will cause an error.

# # Linked issue(s)/Pull request(s)

#109 
#110 
#141 

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [x] Doc update
- [ ] Docker update

## Related Component

- [x] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [x] Windows
  - [x] Mac OS
  - [x] Linux
- Python version:
  - [x] 3.6
  - [x] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [x] New release package
- [x] New docker image

## Checklist

- [x] Add/update the related comments
- [x] Add/update the related tests
- [x] Add/update the related documentations
- [x] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/backends/raw/node.cpp [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+        return "Append, clear and resize function only support for list attribute.";
+      }
+
+      const char* MaxSlotNumberError::what() const noexcept
+      {
+        return "Reach the max number of slot.";
+      }
+}
+  }
+}
```
Comment (Jinyu-W) [inline maro/backends/raw/node.cpp:618]: a new line is needed at the end of each file

File: maro/backends/raw/frame.cpp [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+        return "Not exist attribute type.";
+      }
+
+      const char* FrameInvalidNodeNumerError::what() const noexcept
+      {
+        return "Node number must be greater than 0.";
+      }
+    }
+  }
+}
```
Comment (Jinyu-W) [inline maro/backends/raw/frame.cpp:393]: a new line is needed at the end of each file

File: maro/backends/raw/snapshotlist.cpp [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+        return "List attribute querying need at lease one snapshot, it does not support invalid tick padding.";
+      }
+
+      const char* SnapshotListQueryNoNodeIndexError::what() const noexcept
+      {
+        return "List attribute querying need one alive node index.";
+      }
+    }
+  }
+}
```
Comment (Jinyu-W) [inline maro/backends/raw/snapshotlist.cpp:533]: a new line is needed at the end of each file

**Review by Jinyu-W — CHANGES_REQUESTED** [review]

File: docs/source/key_components/data_model.rst [hunk 4]
```diff
@@ -139,19 +155,34 @@ All supported data types for the attribute of the node:
...
      - Range
-   * - i2
-     - int16_t
+   * - Attribute.Byte
+     - char
+     - -128 .. 127
+   * - Attribute.UByte
+     - unsigned char
+     - 0 .. 255
+   * - Attribute.Short (i2)
```
[the remark this answers is not in the document]
Comment (chaosddp) [reply docs/source/key_components/data_model.rst:164]: yes, used to compatible with old implementation

File: maro/backends/_raw_backend_attr_acc_.pyx.tml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+        self._backend._frame.set_value[{T}](node_index, self._attr_type, slot_index, value)
+
+    cdef object get_value(self, NODE_INDEX node_index, SLOT_INDEX slot_index) except +:
+        return self._backend._frame.get_value[{T}](node_index, self._attr_type, slot_index)
+
+    cdef void append_value(self, NODE_INDEX node_index, object value) except +:
+        self._backend._frame.append_to_list[{T}](node_index, self._attr_type, value)
+
+    cdef void insert_value(self, NODE_INDEX node_index, SLOT_INDEX slot_index, object value) except +:
+        self._backend._frame.insert_to_list[{T}](node_index, self._attr_type, slot_index, value)
```
[the remark this answers is not in the document]
Comment (chaosddp) [reply maro/backends/_raw_backend_attr_acc_.pyx.tml:14]: this is the class template, a new line will be inserted in the generated code

**Review by Jinyu-W — APPROVED** [review]

**Pull request merged**